File: packages/llm-browser/src/llm_browser/js/survey.js
// What the page offers an author before any selector is written: the elements
// that carry a name, the links that share a shape, the structures that repeat,
// and how long the page has been up. Reads only — never clicks, never scrolls.
//
// Raw material, not the answer: which landmark outranks which, what a link
// shape is called and which repeats are the same card live in
// `llm_browser.survey`, so they can be tested without a browser.
(el) => {
  const limits = SURVEY_LIMITS_JSON;
  const doc = el.ownerDocument;
  const squeeze = (value, max) =>
    String(value || "")
      .replace(/\s+/g, " ")
      .trim()
      .slice(0, max);
  const classesOf = (node) => Array.from(node.classList);

  const testidOf = (node) => {
    for (const attribute of limits.testid_attributes) {
      const value = node.getAttribute(attribute);
      if (value) return { testid_attribute: attribute, testid: value };
    }
    return { testid_attribute: null, testid: null };
  };

  // True when a raw cap cut a list short: the answer is a sample of the page
  // rather than the page, and `Survey.truncated` says so.
  let truncated = false;

  const landmarks = [];
  for (const node of doc.querySelectorAll("*")) {
    if (landmarks.length >= limits.max_raw_landmarks) {
      truncated = true;
      break;
    }
    const { testid_attribute, testid } = testidOf(node);
    const aria_label = node.getAttribute("aria-label");
    const id = node.getAttribute("id");
    const role = node.getAttribute("role");
    if (!testid && !aria_label && !id && !role) continue;
    landmarks.push({
      tag: node.tagName.toLowerCase(),
      testid_attribute,
      testid,
      aria_label,
      id,
      role,
      text: squeeze(node.innerText || node.textContent, limits.text_max),
    });
  }

  const anchors = Array.from(doc.querySelectorAll("a[href]"));
  if (anchors.length > limits.max_hrefs) truncated = true;
  const hrefs = anchors
    .slice(0, limits.max_hrefs)
    .map((anchor) => anchor.getAttribute("href"));

  // A repeat is a run of siblings of the same tag: the cards, rows and list
  // items a page is built of. Their classes come along so the caller can tell
  // one component's cards from the next one's.
  // Grouped by tag *and* class, or a table of thirty story rows mixed in with
  // its spacers reads as ninety-six `tr` rather than thirty of one kind. The
  // build's numbering comes off first, so cards the bundler numbered
  // one-by-one still group together.
  const suffix = new RegExp(limits.class_suffix);
  const kindOf = (node) =>
    node.tagName.toLowerCase() +
    "|" +
    classesOf(node)
      .map((token) => token.replace(suffix, "") || token)
      .sort()
      .join(" ");

  const repeats = [];
  for (const parent of doc.querySelectorAll("*")) {
    if (repeats.length >= limits.max_raw_repeats) {
      truncated = true;
      break;
    }
    const children = Array.from(parent.children);
    if (children.length < limits.min_siblings) continue;
    const byKind = new Map();
    for (const child of children) {
      const kind = kindOf(child);
      if (!byKind.has(kind)) byKind.set(kind, []);
      byKind.get(kind).push(child);
    }
    const parentTestid = testidOf(parent);
    for (const members of byKind.values()) {
      const tag = members[0].tagName.toLowerCase();
      if (members.length < limits.min_siblings) continue;
      const shared = classesOf(members[0]).filter((token) =>
        members.every((member) => member.classList.contains(token)),
      );
      repeats.push({
        tag,
        count: members.length,
        classes: classesOf(members[0]),
        shared_classes: shared,
        parent: {
          tag: parent.tagName.toLowerCase(),
          testid_attribute: parentTestid.testid_attribute,
          testid: parentTestid.testid,
          id: parent.getAttribute("id"),
        },
        nested_controls: Array.from(
          members[0].querySelectorAll("button, a, input"),
        )
          .slice(0, limits.max_nested_controls)
          .map((control) => ({
            tag: control.tagName.toLowerCase(),
            text: squeeze(
              control.innerText || control.value || control.textContent,
              limits.nested_text_max,
            ),
          })),
      });
    }
  }

  return {
    title: squeeze(doc.title, limits.text_max),
    url: doc.location.href,
    ready_state: doc.readyState,
    since_navigation_ms: Math.round(performance.now()),
    landmarks,
    hrefs,
    repeats,
    truncated,
  };
}

File: packages/llm-browser/src/llm_browser/survey.py
"""``survey``: what a page is made of, in one read and one page of output.

The page hands over raw material — elements that carry a name, every href,
every run of siblings — and these rules turn it into the three lists an author
writes steps from: landmarks, link shapes and repeats. Every list is capped
here, so a page with ten thousand elements answers in the same breath as a
page with ten.
"""

import re
from collections import Counter
from typing import TYPE_CHECKING

from llm_browser import constants
from llm_browser.explore_selectors import (
    css_quoted,
    escaped_id,
    generated_id,
    hashed_class,
    href_prefix,
)
from llm_browser.scripts import survey_js
from llm_browser.survey_models import (
    Hydration,
    Landmark,
    LinkShape,
    Repeat,
    Survey,
    SurveyNodeRead,
    SurveyRead,
    SurveyRepeatRead,
)

if TYPE_CHECKING:
    from llm_browser.session import BrowserSession

# A build's numbering on the end of a class name: `card-0-2-3`, `title-17`,
# `css-1x2y3z`. What is left is what the next deploy will still call it.
CLASS_SUFFIX = re.compile(r"(?:-(?:\d+|[A-Za-z0-9]*\d[A-Za-z0-9]*))+$")

# Hrefs that go nowhere a step could follow.
DEAD_HREF_SCHEMES = ("#", "javascript:", "mailto:", "tel:")


def class_stem(token: str) -> str:
    """A class name without the build's numbering on the end."""
    return CLASS_SUFFIX.sub("", token) or token


def landmark_selector(node: SurveyNodeRead) -> tuple[int, str] | None:
    """The sturdiest name the element answers to, with its rank: a test id
    outranks a label, a label an id, an id a role. ``None`` for an element
    that offers none of them.

    Same order as ``explore``'s candidates, minus what CSS cannot say: a role
    selects by role alone, and the accessible name stays in ``text``.
    """
    if node.testid and node.testid_attribute:
        return 0, f"[{node.testid_attribute}={css_quoted(node.testid)}]"
    if node.aria_label:
        return 1, f"[aria-label={css_quoted(node.aria_label)}]"
    if node.id and not generated_id(node.id):
        return 2, f"#{escaped_id(node.id)}"
    if node.role:
        return 3, f"[role={css_quoted(node.role)}]"
    return None


def landmarks_of(nodes: list[SurveyNodeRead], max_items: int) -> list[Landmark]:
    """The named elements, deduped by selector and best first.

    Document order inside a rank: the header's landmarks read before the
    footer's, which is the order an author reads the page in anyway.
    """
    found: dict[str, Landmark] = {}
    ranks: dict[str, int] = {}
    for node in nodes:
        named = landmark_selector(node)
        if named is None:
            continue
        rank, selector = named
        seen = found.get(selector)
        if seen is None:
            found[selector] = Landmark(
                selector=selector, tag=node.tag, text=node.text, count=1
            )
            ranks[selector] = rank
        else:
            seen.count += 1
    ranked = sorted(found.values(), key=lambda mark: ranks[mark.selector])
    return ranked[:max_items]


def link_shape(href: str) -> tuple[str, str] | None:
    """A link's family as ``(shape, selector)``; ``None`` when it has none.

    ``/mission/4821`` and ``/mission/5109`` are one shape, ``/mission/<id>``;
    so are ``item?id=1`` and ``item?id=2``, as ``item?<query>``.
    """
    if not href or href.startswith(DEAD_HREF_SCHEMES):
        return None
    path = href.split("?")[0].split("#")[0]
    prefix = href_prefix(href)
    if prefix is None:
        stem, mark = path, ""
    elif prefix.endswith("/"):
        stem, mark = prefix, "<id>"
    else:
        stem, mark = prefix, "?<query>" if "?" in href else ""
    if not stem:
        return None
    return f"{stem}{mark}", f"a[href^={css_quoted(stem)}]"


def link_shapes_of(hrefs: list[str], max_shapes: int) -> list[LinkShape]:
    """The link families the page navigates by, the busiest first."""
    shapes = [shape for shape in map(link_shape, hrefs) if shape is not None]
    counts = Counter(shapes)
    return [
        LinkShape(shape=shape, selector=selector, count=count)
        for (shape, selector), count in counts.most_common(max_shapes)
    ]


def repeat_selector(repeat: SurveyRepeatRead) -> str:
    """What to write against every member of the run.

    A class every member carries is the whole selector, preferring one the
    build did not number: `dense` outlives `sc-card-0-2-1`. Failing that the
    run is named by what holds it, which is why the parent's test id is read.
    """
    stable = [
        token
        for token in repeat.shared_classes
        if class_stem(token) == token and not hashed_class(token)
    ]
    named = stable or repeat.shared_classes
    if named:
        return f"{repeat.tag}.{escaped_id(named[0])}"
    parent = repeat.parent
    if parent.testid and parent.testid_attribute:
        return f"[{parent.testid_attribute}={css_quoted(parent.testid)}] > {repeat.tag}"
    if parent.id and not generated_id(parent.id):
        return f"#{escaped_id(parent.id)} > {repeat.tag}"
    return f"{parent.tag} > {repeat.tag}"


def repeats_of(runs: list[SurveyRepeatRead], max_repeats: int) -> list[Repeat]:
    """The repeated structures, the busiest first.

    Two grids of the same card are one card: they answer to the same selector,
    so their counts are the same number — which is the number an author is
    about to write a `read` against.
    """
    merged: dict[str, Repeat] = {}
    for run in runs:
        selector = repeat_selector(run)
        seen = merged.get(selector)
        if seen is None:
            merged[selector] = Repeat(
                selector=selector,
                count=run.count,
                nested_controls=run.nested_controls,
            )
        else:
            seen.count += run.count
    ranked = sorted(merged.values(), key=lambda run: -run.count)
    return ranked[:max_repeats]


def survey_of(read: SurveyRead, max_items: int) -> Survey:
    """The rules, over what the page handed back — no browser involved."""
    return Survey(
        title=read.title,
        url=read.url,
        hydration=Hydration(
            since_navigation_ms=read.since_navigation_ms,
            ready_state=read.ready_state,
        ),
        landmarks=landmarks_of(read.landmarks, max_items),
        link_shapes=link_shapes_of(read.hrefs, constants.SURVEY_MAX_LINK_SHAPES),
        repeats=repeats_of(read.repeats, constants.SURVEY_MAX_REPEATS),
    )


def survey(
    session: "BrowserSession", max_items: int = constants.SURVEY_MAX_ITEMS
) -> Survey:
    """What the page offers a step, in one read: the first call of a session.

    The named elements to write selectors against, the link families the page
    navigates by, the structures it repeats — the cards, rows and items — and
    how long it had been up when asked. Nothing is clicked and nothing is
    scrolled, so the answer is the page as it stands.

    Where ``explore`` answers "is this selector right", ``survey`` answers the
    question before it: which selectors are there to try.
    """
    read = SurveyRead.model_validate(session.evaluate_document(survey_js()))
    return survey_of(read, max_items)
